bokehjs/src/lib/models/widgets/file_input.ts

```diff
@@ -16,43 +16,59 @@ export class FileInputView extends WidgetView {
     if (this.dialogEl == null) {
       this.dialogEl = document.createElement('input')
       this.dialogEl.type = "file"
-      this.dialogEl.multiple = false
-      this.dialogEl.onchange = (e) => this.load_file(e)
+      this.dialogEl.multiple = this.model.multiple
+      this.dialogEl.onchange = async (e) => await this.load_files(e)
       this.el.appendChild(this.dialogEl)
     }
     if (this.model.accept != null && this.model.accept != '')
       this.dialogEl.accept = this.model.accept
+
     this.dialogEl.style.width = `{this.model.width}px`
     this.dialogEl.disabled = this.model.disabled
   }
 
-  load_file(e: any): void {
-    const reader = new FileReader()
-    this.model.filename = e.target.files[0].name
-    reader.onload = (e) => this.file(e)
-    reader.readAsDataURL(e.target.files[0])
+  async load_files(e: any): Promise<void> {
+    const files = e.target.files
+    const value: string[] = []
+    const filename: string[] = []
+    const mime_type: string[] = []
+    let i: number
+
+    for (i = 0; i < files.length; i++){
+      filename.push(files[i].name)
+      const data_url = await this.readfile(files[i])
+      const [, mime, , data] = data_url.split(/[:;,]/, 4)
+      value.push(data)
+      mime_type.push(mime)
+    }
+    if (this.model.multiple) {
+      this.model.filename = filename
+      this.model.mime_type = mime_type
+      this.model.value = value
+    } else {
+      this.model.filename = filename[0]
+      this.model.mime_type = mime_type[0]
+      this.model.value = value[0]
+    }
   }
 
-  file(e: any): void {
-    const file = e.target.result
-    const file_arr = file.split(",")
-
-    const content = file_arr[1]
-    const header = file_arr[0].split(":")[1].split(";")[0]
-
-    this.model.value = content
-    this.model.mime_type = header
-
+  readfile(file: any): Promise<string> {
+    return new Promise<any>((resolve) => {
```

Comment:
I will leave this one as an exercise. Let me know if you have any issues/questions with this one.

Reply:
My solution below gets rid of "any"-type but needs a String() cast in order to satisfy the checks of the compiler. Although _e.target.result_ should contain a string because of _readAsDataURL()_ the compiler complains about incompatible types if I remove the cast (???). Apparently there are things that I still do not understand.
```
  readfile(file: File): Promise<string> {
    return new Promise<string>((resolve) => {
      const reader = new FileReader()
      reader.onload = (e) => {
        if (e.target) {
            resolve(e.target.result ? String(e.target.result) : "")
        }
        else {resolve("")}
      }
      reader.readAsDataURL(file)
    })
  }
```